internal/transform/scope-html.go

```diff
@@ -95,72 +95,81 @@ func injectDefineVars(n *astro.Node, values []string) {
+						return
+					}
+				case astro.EmptyAttribute:
+					// instead of an empty string
+					attr.Type = astro.QuotedAttribute
+					attr.Val = scopedClass
+					n.Attr[i] = attr
+					return
+				case astro.QuotedAttribute, astro.TemplateLiteralAttribute:
+					// as a plain string
+					attr.Val = fmt.Sprintf(`%s %s`, attr.Val, scopedClass)
+					n.Attr[i] = attr
+					return
+				case astro.ExpressionAttribute:
+					// as an expression
+					attr.Val = fmt.Sprintf(`(%s) + " %s"`, attr.Val, scopedClass)
 					n.Attr[i] = attr
 					return
 				}
-			case astro.EmptyAttribute:
-				// instead of an empty string
-				attr.Type = astro.QuotedAttribute
-				attr.Val = scopedClass
-				n.Attr[i] = attr
-				return
-			case astro.QuotedAttribute, astro.TemplateLiteralAttribute:
-				// as a plain string
-				attr.Val = fmt.Sprintf(`%s %s`, attr.Val, scopedClass)
-				n.Attr[i] = attr
-				return
-			case astro.ExpressionAttribute:
-				// as an expression
-				attr.Val = fmt.Sprintf(`(%s) + " %s"`, attr.Val, scopedClass)
-				n.Attr[i] = attr
-				return
 			}
-		}
 
-		if attr.Key == "class:list" {
-			switch attr.Type {
-			case astro.EmptyAttribute:
-				// instead of an empty string
-				attr.Type = astro.QuotedAttribute
-				attr.Val = "astro-" + opts.Scope
-				n.Attr[i] = attr
-				return
-			case astro.QuotedAttribute, astro.TemplateLiteralAttribute:
-				// as a plain string
-				attr.Val = attr.Val + " astro-" + opts.Scope
-				n.Attr[i] = attr
-				return
-			case astro.ExpressionAttribute:
-				// as an expression
-				attr.Val = fmt.Sprintf(`[(%s), "%s"]`, attr.Val, scopedClass)
-				n.Attr[i] = attr
-				return
+			if attr.Key == "class:list" {
+				switch attr.Type {
+				case astro.EmptyAttribute:
+					// instead of an empty string
+					attr.Type = astro.QuotedAttribute
+					attr.Val = "astro-" + opts.Scope
+					n.Attr[i] = attr
+					return
+				case astro.QuotedAttribute, astro.TemplateLiteralAttribute:
+					// as a plain string
+					attr.Val = attr.Val + " astro-" + opts.Scope
+					n.Attr[i] = attr
+					return
+				case astro.ExpressionAttribute:
+					// as an expression
+					attr.Val = fmt.Sprintf(`[(%s), "%s"]`, attr.Val, scopedClass)
+					n.Attr[i] = attr
+					return
+				}
 			}
 		}
+
+		// If there's a spread attribute, `class` might be there, so do not inject `class` here
+		// `class` will be injected by the `spreadAttributes` helper
+		if hasSpreadAttr {
+			return
+		}
+		n.Attr = append(n.Attr, astro.Attribute{
+			Key:  "class",
+			Type: astro.QuotedAttribute,
+			Val:  scopedClass,
+		})
+	} else {
+		n.Attr = append(n.Attr, astro.Attribute{
+			Key:  fmt.Sprintf(`data-astro-hash-%s`, opts.Scope),
+			Type: astro.EmptyAttribute,
+		})
 	}
-	// If there's a spread attribute, `class` might be there, so do not inject `class` here
-	// `class` will be injected by the `spreadAttributes` helper
-	if hasSpreadAttr {
-		return
-	}
-	// If we didn't find an existing class attribute, let's add one
```

Comment:
Not sure if it's a mistake this comment was removed, but it looks like a nice one to preserve.

Reply:
Ah shoot, definitely not my intention. Great catch! Thank you!